qbft/spectest/tests/controller/latemsg/late_commit_past_instance.go

```diff
@@ -36,30 +39,81 @@ func LateCommitPastInstance() tests.SpecTest {
 				DecidedVal: testingutils.TestingQBFTFullData,
 				DecidedCnt: 1,
 			},
-			ControllerPostRoot: postRoot,
+			ControllerPostRoot:  sc.Root(),
+			ControllerPostState: sc.ExpectedState,
 		}
 	}
 
+	lateMsg := testingutils.TestingCommitMultiSignerMessageWithHeight([]*bls.SecretKey{ks.Shares[4]}, []types.OperatorID{4}, 4)
+	sc := lateCommitPastInstanceStateComparison(6, lateMsg)
+
 	return &tests.ControllerSpecTest{
 		Name: "late commit past instance",
 		RunInstanceData: []*tests.RunInstanceData{
-			instanceData(qbft.FirstHeight, "24cf697092529cfab3ab06b969d8696692c8bcbb9f41a954f71dc74c3b1d7e97"),
-			instanceData(1, "8aa5464b119518f178d81edf4cea1f4c918f9e084e5262a0e276d3afb00ba620"),
-			instanceData(2, "1799fe0981ae08bde1eae9fef88ef8035f5952974647786287a1a8c36544a5da"),
-			instanceData(3, "2ddf2b8c2f35c8115ddd68120c71e64809bfea6b023ed13e177c2474a95d137d"),
-			instanceData(4, "318816cc8819ad062996704fb4b9990b8088ade0cb4c26816ea0965783bff12e"),
-			instanceData(5, "fd83cdee705de628cf4a9baf9e662c424bf0942c63a680266d7db872d16e9f0a"),
+			instanceData(qbft.FirstHeight),
+			instanceData(1),
+			instanceData(2),
+			instanceData(3),
+			instanceData(4),
+			instanceData(5),
 			{
 				InputValue: []byte{1, 2, 3, 4},
 				InputMessages: []*qbft.SignedMessage{
-					testingutils.TestingCommitMultiSignerMessageWithHeight(
-						[]*bls.SecretKey{ks.Shares[4]},
-						[]types.OperatorID{4},
-						4,
-					),
+					lateMsg,
 				},
-				ControllerPostRoot: "691ac09194b0f6c67e85de702293f19863b0f5ee19e17c3764e411cee253f8b7",
+				ControllerPostRoot:  sc.Root(),
+				ControllerPostState: sc.ExpectedState,
 			},
 		},
 	}
 }
+
+func lateCommitPastInstanceStateComparison(height qbft.Height, lateMsg *qbft.SignedMessage) *comparable.StateComparison {
+	ks := testingutils.Testing4SharesSet()
+	allMsgs := testingutils.ExpectedDecidingMsgsForHeightWithRoot(testingutils.TestingQBFTRootData, testingutils.TestingQBFTFullData, testingutils.TestingIdentifier, 5, ks)
+	offset := 7 // 7 messages per height (1 propose + 3 prepare + 3 commit)
+
+	contr := testingutils.NewTestingQBFTController(
+		testingutils.TestingIdentifier,
+		testingutils.TestingShare(testingutils.Testing4SharesSet()),
+		testingutils.TestingConfig(testingutils.Testing4SharesSet()),
+	)
+
+	for i := 0; i <= int(height); i++ {
+		contr.Height = qbft.Height(i)
+		msgs := allMsgs[offset*i : offset*(i+1)]
+
+		instance := &qbft.Instance{
+			StartValue: []byte{1, 2, 3, 4},
+			State: &qbft.State{
+				Share:  testingutils.TestingShare(testingutils.Testing4SharesSet()),
+				ID:     testingutils.TestingIdentifier,
+				Round:  qbft.FirstRound,
+				Height: qbft.Height(i),
+			},
+		}
+
+		// last height
+		if lateMsg != nil && qbft.Height(i) == height {
+			comparable.SetSignedMessages(instance, []*qbft.SignedMessage{})
```

Comment:
why is this an empty msg?

Reply:
It isn't an empty message. It is a slice that contains no messages.

Note that the last `instanceData` we are creating is the "lateMsg" one. This means it has a previously seen height (4). The last height check ensures we indeed create the instance according to the lateMsg's data and not according to the expected height.

Reply:
refactored to have less instanceDatas.

Reply:
add some explanation of what it does.. super hard to understand

Reply:
changed to `InitContainers()`
There is also a function comment easily viewable with an IDE.